### 3.1 Contents

### 3.2 Introduction

### 3.3 Imports

In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.preprocessing import scale

### 3.4 Load the data

#### 3.4.1 Cyclistic DataFrame - combined, clean

In [3]:
data = pd.read_csv('df_cyclistic_clean.csv')

In [4]:
#validate dtypes are correct
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5675095 entries, 0 to 5675094
Data columns (total 11 columns):
 #   Column         Dtype  
---  ------         -----  
 0   ride_id        object 
 1   rideable_type  object 
 2   started_at     object 
 3   ended_at       object 
 4   start_lat      float64
 5   start_lng      float64
 6   end_lat        float64
 7   end_lng        float64
 8   member_casual  object 
 9   ride_length    object 
 10  day_of_week    object 
dtypes: float64(4), object(7)
memory usage: 476.3+ MB


In [8]:
#dtype did not sustain through filesave. convert 'ride_length' to 
data['ride_length'] = pd.to_timedelta(data['ride_length'])

In [10]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5675095 entries, 0 to 5675094
Data columns (total 11 columns):
 #   Column         Dtype          
---  ------         -----          
 0   ride_id        object         
 1   rideable_type  object         
 2   started_at     object         
 3   ended_at       object         
 4   start_lat      float64        
 5   start_lng      float64        
 6   end_lat        float64        
 7   end_lng        float64        
 8   member_casual  object         
 9   ride_length    timedelta64[ns]
 10  day_of_week    object         
dtypes: float64(4), object(6), timedelta64[ns](1)
memory usage: 476.3+ MB


In [11]:
#validate columns are correct
data.columns

Index(['ride_id', 'rideable_type', 'started_at', 'ended_at', 'start_lat',
       'start_lng', 'end_lat', 'end_lng', 'member_casual', 'ride_length',
       'day_of_week'],
      dtype='object')

In [12]:
#validate shape is consistent
data.shape

(5675095, 11)

In [13]:
# check out first 5 rows 
data.head()

ride_id  rideable_type           started_at             ended_at  \
0  C1D650626C8C899A  electric_bike  2024-01-12 15:30:27  2024-01-12 15:37:59   
1  EECD38BDB25BFCB0  electric_bike  2024-01-08 15:45:46  2024-01-08 15:52:59   
2  F4A9CE78061F17F7  electric_bike  2024-01-27 12:27:19  2024-01-27 12:35:19   
3  0A0D9E15EE50B171   classic_bike  2024-01-29 16:26:17  2024-01-29 16:56:06   
4  33FFC9805E3EFF9A   classic_bike  2024-01-31 05:43:23  2024-01-31 06:09:35   

   start_lat  start_lng    end_lat    end_lng member_casual     ride_length  \
0  41.903267 -87.634737  41.889177 -87.638506        member 0 days 00:07:32   
1  41.902937 -87.634440  41.889177 -87.638506        member 0 days 00:07:13   
2  41.902951 -87.634470  41.889177 -87.638506        member 0 days 00:08:00   
3  41.884295 -87.633963  41.921822 -87.644140        member 0 days 00:29:49   
4  41.948797 -87.675278  41.889177 -87.638506        member 0 days 00:26:12   

  day_of_week  
0      Friday  
1      Monday  
2    Saturday  
3      Monday  
4   Wednesday

### 3.5 Explore the Data

### 3.5.1 create sample set to do EDA calculations on

In [14]:
data.describe()

start_lat     start_lng       end_lat       end_lng  \
count  5.675095e+06  5.675095e+06  5.675095e+06  5.675095e+06   
mean   4.190229e+01 -8.764621e+01  4.190267e+01 -8.764642e+01   
std    4.476176e-02  2.749779e-02  5.631905e-02  1.153537e-01   
min    4.164000e+01 -8.791000e+01  1.606000e+01 -1.440500e+02   
25%    4.188096e+01 -8.766000e+01  4.188096e+01 -8.766000e+01   
50%    4.189745e+01 -8.764182e+01  4.189776e+01 -8.764288e+01   
75%    4.193000e+01 -8.762943e+01  4.193000e+01 -8.762952e+01   
max    4.207000e+01 -8.752000e+01  8.796000e+01  1.525300e+02   

                     ride_length  
count                    5675095  
mean   0 days 00:15:37.154399652  
std    0 days 00:33:02.265702909  
min            -2 days +02:11:41  
25%              0 days 00:05:35  
50%       0 days 00:09:47.348000  
75%       0 days 00:17:22.361500  
max              1 days 01:09:22

If dataframe is too large know that you can sample (ended up not being necessary here)

In [20]:
# creating sample set (did not end up using)
# sample = data.sample(frac=0.1)  # Sample 10% of the data
# sample.describe()

### 3.5.2 Calculations to better understand data
#### mean ride_length, max ride_length, mode of day of week

Open your spreadsheet application, then complete the following steps:
1. Where relevant, make columns consistent and combine them into a single worksheet.
2. Clean and transform your data to prepare for analysis.
3. Conduct descriptive analysis.
4. Run a few calculations in one file to get a better sense of the data layout. Options:
    ● Calculate the mean of ride_length
    ● Calculate the max ride_length
    ● Calculate the mode of day_of_week
5. Create a pivot table to quickly calculate and visualize the data. Options:
    ● Calculate the average ride_length for members and casual riders. Try rows = member_casual; Values = Average of
    ride_length.
    ● Calculate the average ride_length for users by day_of_week. Try columns = day_of_week; Rows = member_casual;
    Values = Average of ride_length.
    ● Calculate the number of rides for users by day_of_week by adding Count of trip_id to Values.
6. Open another file and perform the same descriptive analysis steps. Explore different seasons to make some initial
observations.
7. Once you have spent some time working with the individual spreadsheets, merge them into a full-year view. Do this with the
tool you have chosen to use to perform your

Follow these steps for using R
Open R Studio and use this script to complete the following steps:
1. Import your data.
2. Make columns consistent and merge them into a single dataframe.
3. Clean up and add data to prepare for analysis.
4. Conduct descriptive analysis.
5. Export a summary file for further analysis.

In [15]:
mean_ride_length = data['ride_length'].mean()
print(f"Mean ride length: {mean_ride_length}")

Mean ride length: 0 days 00:15:37.154399652


In [16]:
max_ride_length = data['ride_length'].max()
print(f"Max ride length: {max_ride_length}")

Max ride length: 1 days 01:09:22


In [17]:
min_ride_length = data['ride_length'].min()
print(f"Min ride length: {min_ride_length}")

Min ride length: -2 days +02:11:41


'ride_length' values should be absolute values (not negative). count negative values, investigate and convert values to absolute values

In [26]:
# count number of negative values in 'ride_length'

negative_ride_length_count = (data['ride_length'] < pd.Timedelta(0)).sum()
print(negative_ride_length_count)

227


In [37]:
# create dataframe of rows with negative values in 'ride_length' column
negative_ride_length_df = data[data['ride_length'] < pd.Timedelta(0)]

In [38]:
# show negative_ride_length_df (formatted for easier viewing)
selected_columns = [0, 2, 3, 4, 5, 6, 7, 8, 9, 10]
print(negative_ride_length_df.iloc[:, selected_columns])

                  ride_id               started_at                 ended_at  \
19682    3178F8568B02A678      2024-01-10 07:26:59      2024-01-10 07:26:58   
87139    15792904345CE6AE      2024-01-08 08:21:34      2024-01-08 08:21:33   
90511    7C2B9266971D5862      2024-01-21 14:57:56      2024-01-21 14:21:04   
90598    3ADCE21FEE848DAC      2024-01-18 09:59:36      2024-01-18 09:19:38   
94521    D688B0ED24D11C65      2024-01-21 17:22:48      2024-01-21 16:44:58   
...                   ...                      ...                      ...   
5643847  3A587733462DA2C3  2024-11-03 01:57:49.758  2024-11-03 01:17:29.247   
5649768  4BD1F0EE07E79421  2024-11-03 01:51:30.818  2024-11-03 01:16:39.875   
5656455  C497066FD1D05EEE  2024-11-03 01:58:02.303  2024-11-03 01:03:32.261   
5656670  6D8CC446C1251515  2024-11-03 01:52:30.180  2024-11-03 01:06:44.950   
5673682  E8F98270528A59CC  2024-11-03 01:54:55.644  2024-11-03 01:03:12.879   

         start_lat  start_lng    end_lat    end_lng

In [41]:
percent_negative_ride_length = len(negative_ride_length_df)/len(data)
print(f"% of data with negative values in 'ride_length' column: {percent_negative_ride_length}")

% of data with negative values in 'ride_length' column: 3.999933040768481e-05


I've identified that there are 227 rows with negative values in ride_length column. They do not have any obvious patterns. I have 3 options for error handling:

a) remove rows

b) correct timestamps

c) flag rows as invalid

For this project I am choosing to remove the rows. 

My first thought was that these negative values were due to the rides starting around midnight and going into the next day. However. because I did the calculation with datetime dtype this should not be an issue. 

I then thought that the started_at and ended_at entries, could have been switched, but due to the nature of this being a case study, I am not able to confirm this. I don't want to imput incorrect data. I'm not entirely sure if this is the case and if so, what does that mean for the started_at/ended_at stations? Should they also be switched.

These 227 rows make up 0.00004% of the dataset, so I'm deciding to simply remove the rows.

In [47]:
# remove rows where 'started_at' is after 'ended_at'
data = data[data['started_at'] <= data['ended_at']]

In [48]:
# check new dataframe (previously 5675095 entries)
print(len(data))

5674868


In [52]:
print(5675095 - len(data))

227


In [49]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5674868 entries, 0 to 5675094
Data columns (total 11 columns):
 #   Column         Dtype          
---  ------         -----          
 0   ride_id        object         
 1   rideable_type  object         
 2   started_at     object         
 3   ended_at       object         
 4   start_lat      float64        
 5   start_lng      float64        
 6   end_lat        float64        
 7   end_lng        float64        
 8   member_casual  object         
 9   ride_length    timedelta64[ns]
 10  day_of_week    object         
dtypes: float64(4), object(6), timedelta64[ns](1)
memory usage: 519.5+ MB


In [54]:
data.describe()

start_lat     start_lng       end_lat       end_lng  \
count  5.674868e+06  5.674868e+06  5.674868e+06  5.674868e+06   
mean   4.190228e+01 -8.764621e+01  4.190267e+01 -8.764642e+01   
std    4.476135e-02  2.749757e-02  5.631910e-02  1.153557e-01   
min    4.164000e+01 -8.791000e+01  1.606000e+01 -1.440500e+02   
25%    4.188096e+01 -8.766000e+01  4.188096e+01 -8.766000e+01   
50%    4.189745e+01 -8.764182e+01  4.189776e+01 -8.764288e+01   
75%    4.193000e+01 -8.762943e+01  4.193000e+01 -8.762952e+01   
max    4.207000e+01 -8.752000e+01  8.796000e+01  1.525300e+02   

                     ride_length  
count                    5674868  
mean   0 days 00:15:37.277931443  
std    0 days 00:33:00.744531268  
min              0 days 00:00:00  
25%              0 days 00:05:35  
50%       0 days 00:09:47.377000  
75%       0 days 00:17:22.408500  
max              1 days 01:09:22

In [53]:
mode_ride_length = data['day_of_week'].mode()
print(f"Mode ride length: {mode_ride_length}")

Mode ride length: 0    Saturday
Name: day_of_week, dtype: object


### 3.5.3 Calculations by rider type

#### 3.5.3.1 average ride length by rider type

In [57]:
# calculate average 'ride_length' for each rider type
avg_ride_length_by_rider_type = data.groupby('member_casual')['ride_length'].mean() 

In [58]:
# print results
print("Average Ride Length by Member Type:")
print(avg_ride_length_by_rider_type)

Average Ride Length by Member Type:
member_casual
casual   0 days 00:21:14.557806662
member   0 days 00:12:18.036518574
Name: ride_length, dtype: timedelta64[ns]


Members average ride length is roughly half the time of a casual rider. Members avg ride length is ~12 min. Casual riders avg ride length is ~21 min. 

In [64]:
# calculate avg ride_length by day_of_week
avg_ride_length_by_day = data.groupby(['day_of_week', 'member_casual'])['ride_length'].mean()

In [69]:
# pivot the data for easier readability
avg_ride_length_table = avg_ride_length_by_day.unstack(level='member_casual')

In [70]:
# print results
print("Average ride length by day of week for each member type:")
print(avg_ride_length_table)

Average ride length by day of week for each member type:
member_casual                    casual                    member
day_of_week                                                      
Friday        0 days 00:20:32.172135869 0 days 00:11:58.961497760
Monday        0 days 00:20:36.635393713 0 days 00:11:45.188644547
Saturday      0 days 00:24:02.983070771 0 days 00:13:38.973748322
Sunday        0 days 00:24:31.118988985 0 days 00:13:40.860195310
Thursday      0 days 00:18:27.486079861 0 days 00:11:47.282485723
Tuesday       0 days 00:18:19.265714665 0 days 00:11:46.751105387
Wednesday     0 days 00:18:45.350695445 0 days 00:12:00.460125607


Members average ride length is fairly consistent, just under 12min with slightly longer (13 min) averages on weekends. Casual riders have slightly more variable average ride lengths which are consistently longer than member average ride lengths. Casual riders average ride length is around 18.5-20 min on weekdays. 24 min on weekends.

#### 3.5.3.2 number of rides for members v casual riders by day_of_week

In [72]:
# group by day_of_week and member_casual, then count the number of rides 
ride_counts = data.groupby(['day_of_week', 'member_casual']).size().reset_index(name='ride_count')

In [75]:
# pivot the data for easier readability
ride_counts_table = ride_counts.pivot(index='day_of_week', columns='member_casual', values='ride_count')

In [76]:
# print the results in a clear table format
print("Number of rides by day of week and member type:")
print(ride_counts_table)

Number of rides by day of week and member type:
member_casual  casual  member
day_of_week                  
Friday         310652  508694
Monday         245874  506664
Saturday       437876  464178
Sunday         362538  400634
Thursday       260039  552535
Tuesday        226124  544713
Wednesday      264304  590043
